Add syntax and description to @layer (#15215)
mdn/content: files/en-us/web/css/@layer/index.md

---
{{CSSRef}}

The **`@layer`** [CSS](/en-US/docs/Web/CSS) [at-rule](/en-US/docs/Web/CSS/At-rule) is used to declare a cascade layer and can also used to define the order of precedence in case of multiple cascade layers.

## Syntax

```css
@layer layer-name {rules};
@layer layer-name;
@layer layer-name, layer-name, layer-name;
@layer {rules};

> @import url layer( layer-name );

Reviewer: Should the @import be in the @layer page?
And, if it is, should it come first as @import is supposed to come before, or precede, any other styles.

Author: Removed @import from the Syntax

```

where:
- _layer-name_
  - : Is the name of the cascade layer.
- _rules_
  - : Is the set of CSS rules in the cascade layer.
- _url_
  - : Is a {{CSSxRef("&lt;string&gt;")}}, a `<url>`, or a {{CSSxRef("url")}} function representing the location of the resource to import. The URL may be absolute or relative.

## Description

> Rules within a cascade layer cascade together, giving more control over the cascade to web developers. Any styles not in a layer are gathered together and placed into an anonymous layer that comes after all the declared layers. This means that any styles declared outside of a layer will override styles declared in a layer.

Reviewer: ```suggestion
Rules within a cascade layer cascade together, giving more control over the cascade to web developers. Any styles not in a layer are gathered together and placed into an anonymous layer that comes after all the declared layers. This means that any styles declared outside of a layer will override styles declared in a layer, regardless of specificity.
```

> Rules within a cascade layer cascade together, giving more control over the cascade to web developers. Any styles not in a layer are gathered together and placed into an anonymous layer that comes after all the declared layers. This means that any styles declared outside of a layer will override styles declared in a layer, regardless of specificity.

Reviewer: ```suggestion
Rules within a cascade layer cascade together, giving more control over the cascade to web developers. Any styles not in a layer are gathered together and placed into a single anonymous layer that comes after all the declared layers, named and anonymous. This means that any styles declared outside of a layer will override styles declared in a layer, regardless of specificity.
```

Reviewer: want to make sure people understand there can be multiple anonymous layers.

}
```

The second way is to create a named cascade layer without assigning any styles. This can be a single layer, as shown below:

```css
@layer utilities;
```

> Or multiple layers can be defined at once, as shown below:

Reviewer: ```suggestion
Multiple layers can be defined at once, as shown below:
```

```css
@layer theme, layout, utilities;
```

This is useful because the initial order in which layers are declared indicates which layer has precedence. As with declarations, the last layer to be listed will win if declarations are found in multiple layers. Therefore, with the preceding example, if a competing rule was found in `theme` and `utilities`, the one in `utilities` would win and be applied.

A rule in `utilities` would be applied _even if it has lower specificity_ than the rule in `theme`. This is because once the layer order has been established, specificity and order of appearance are ignored. This enables the creation of simpler CSS selectors because you do not have to ensure that a selector will have high enough specificity to override competing rules; all you need to ensure is that it appears in a later layer.

> **Note:** Having declared your layer names, thus setting their order, you can add CSS rules to the layer by re-declaring the name. The styles are then appended to the layer and the layer order will not be changed.

The third way is to create a cascade layer with no name. For example:

```css
@layer {
  p {
    margin-block: 1rem;
  }
}
```

> This creates an _anonymous cascade layer_. This layer functions in the same way as named layers; however, rules cannot be assigned to it later.

Reviewer: ```suggestion
This creates an _anonymous cascade layer_. This layer functions in the same way as named layers; however, rules cannot be assigned to it later. The order of precedence for anonymous layers is the order in which layers are declared, named or not, and lower than styles declared outside of a layer.
```

> Another way to create a cascade layer is by using {{cssxref("@import")}}. In this case, the rules would be in the imported stylesheet.

Reviewer: should we remind readers that imports have to come first in a stylesheet?

> Another way to create a cascade layer is by using {{cssxref("@import")}}. In this case, the rules would be in the imported stylesheet. Remember that the `@import` at-rule must precede all other types of rules, except the @charset rules.

Reviewer: I think it's enough to keep this and remove import from the syntax box at the start of this document.
